doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,304 @@
+ and check with `VALIDATE CONNECTION` that it's working.
+- User will also need to give the AWS user/role for Materialize the following access to the
+s3 prefixes so that we can write the data.
+  - `s3:ListBucket`: We need this permission to make sure that we are writing to an empty path.
+  - `s3:PutObject`: This is required to actually upload files to S3.
+- If they try to run a `COPY` to S3 without all the permissions, then the operation should fail
+with an error. Note, running a `VALIDATE CONNECTION` does not guarantee that the specific S3
+permissions are set up correctly. These can only be validated when the actual `COPY` command
+is run.
+- If the S3 path given already has some data in it, we should fail the operation.
+- When they run the `COPY` command, to keep things simple we'll block the UI till the operation
+is complete, similar to any other query.
+- If successful, the output of the `COPY` command should return the number of rows written to S3
+similar to a [Postgres `COPY` output](https://www.postgresql.org/docs/current/sql-copy.html) of
+`COPY <row_count>`.
+- If the operation fails midway for some reason or is cancelled by the user, we should show
+a notice to the users that incomplete data could have been written to the S3 path and will
+need to be cleaned up.
+- Since we'll support scalar expressions in the s3 path, user can give a path like
+.`'s3://prefix/path/' || mz_now()` so that they can run the same query again and again without
+needing to update the path or clean up previous data.
+
+## Implementation
+
+### Parser
+#### S3 path
+We should accept the S3 path as an arbitrary scalar expression which we later convert to
+a string in the planner.
+```
+Uri::from_str(lower_uncorrelated().eval().unwrap_str())
+```
+
+#### `MAX FILE SIZE` Option
+We should introduce a `Memory` type which can take either an integer or a parseable string
+like `'2 GB'`, similar to the implementation for `Duration`. The multiplier should be
+1024 to keep it consistent with Postgres. If the value is not provided we should select
+an appropriate default.
+
+Snowflake uses 16MB as default for their `MAX_FILE_SIZE` whereas Redshift defaults to
+6.2GB for similar `MAXFILESIZE` parameter. Also, Snowflake recommends for data in S3 to be
+used in external tables, file sizes should be under 256MB. Given that, 256MB
+can be the default for our `MAX FILE SIZE`.
+
+### Which cluster to run things on?
+Currently existing supported `COPY ... TO STDOUT` command fetches the result to `environmentd`.
+This will not be advisable for `COPY ... TO S3` as the data size could be large. Therefore,
+we should trigger a dataflow in a `clusterd` which should eventually write to S3 using
+user's aws connection details. The next question being, should it be a Storage cluser or a
+Compute cluster (at least till the time the cluster unification is done).
+
+We should use the Compute cluster. Compute already has a concept of
```

Comment:
Please keep in mind that there may be multiple replicas running the same sink, so it'll need to coordinate accordingly. Also, compute doesn't currently assume that replicas run to completion, so the sink must tolerate being stopped and recreated at any time.

An alternative we should consider is that we create an ephemeral materialized view based on the select expression, und do the actual uploading in storage.

Reply:
> Please keep in mind that there may be multiple replicas running the same sink, so it'll need to coordinate accordingly. Also, compute doesn't currently assume that replicas run to completion, so the sink must tolerate being stopped and recreated at any time.

I think it should be fine as long as the computation is perfectly deterministic, right? Then no coordination is required. 

> An alternative we should consider is that we create an ephemeral materialized view based on the select expression, und do the actual uploading in storage.

Are there benefits to this approach? Storage will one day soon support replication.

Reply:
> I think it should be fine as long as the computation is perfectly deterministic, right? Then no coordination is required.

Not if the worker ID is part of the file names. In this case, the output from different replicas is equivalent, but the sharding is different.

> Are there benefits to this approach? Storage will one day soon support replication.

None other than different splitting of concerns. I have no strong feeling here!

Reply:
> Not if the worker ID is part of the file names. In this case, the output from different replicas is equivalent, but the sharding is different.

Oh, that's a really good point. Hrm. We'll need to sort that out.

Reply:
It's not going to be as efficient, but we could start with only a single worker writing to s3.

Also, when a compute sink is killed, there could be multiple multi part uploads in progress. Initially I was thinking these could be kept worker local state, but that wouldn't work if the replicas are themselves killed. What would be a good way to store the progress? Would it need to be similar to a progress collection in storage?

Reply:
There's a certain risk involved with moving everything to a single worker, which is that it might stall out the rest of the system. We should probably discuss alternatives that either don't have this property or avoid it.

Reply:
@benesch and I briefly talked about this, and there seems to be a way out of this:
* Pick a number of parts, and tell the sink how many parts it should have.
* The number could have a default, say 32, or whatever the largest replica is at creation time, or it could be specified by the user.
* Assign records based on a hash function to parts, exchange the data based on the part to workers. We can workshop how this works in detail.
* Upload the individual parts, name them by their part ID.

To produce deterministic outputs, consider:
* adding a read hold for the timestamp to be copied, which corresponds to the as-of of the compute dataflow.
* Then, restarts would use the same timestamp and since it's deterministic, upload the same data again.
* There's probably some work required to teach the adapter about copy commands, and they need to be in the catalog so they can survive envd restarts.

Reply:
Thanks, I will add this to the doc.

Reply:
> The number could have a default, say 32, or whatever the largest replica is at creation time, or it could be specified by the user.

I'd say number of workers in the largest replica at creation time is a good heuristic. I bet the number of cases where someone starts a `COPY` command and then proceeds to change the replica configuration while that is running will be exceedingly rare, so we don't have to optimize for this case.

> There's probably some work required to teach the adapter about copy commands, and they need to be in the catalog so they can survive envd restarts.

Why do we need them to survive envd restarts? I don't think it's clearly specified in this design but AIUI the idea it that a `COPY TO S3` command always only copies a point-in-time snapshot of whatever it reads from. If this is true then it behaves like a `SELECT` in that we can forget about it on client disconnects or envd restarts.

If my understanding is not correct and we do want to support continuous copying of collections to S3 somehow, I think we'd run into problems with determinism. Keep in mind that between envd restarts we could have deployed a new release, which might have changed the optimizer or rendering so that we now produce different outputs than before.

Reply:
> Arguably, one shot operations should be a CREATE operation that creates a named catalog object...

I will add this in future work.